Explainability framework SHAP main code, masking of input Issue

Kathleendeboer opened the pull request:
Code for issue #15 

The code contains:
- Loading the model
- Gathering the data
- Generating tokens representing the input
- Masking the input based on the masked tokens
- Running the model on all combinations of masked input
- Storing the predictions and probabilities

This code still needs: 
- Testing of the inference of the model on the masked combinations 
- Connection to input from Sound and Vision:
   - It now uses the validation mode from the model which needs target labels, maybe add dummy labels?
   - Connect to pre process module (in progress) Issue #2

Kathleendeboer committed:
Explainability framework SHAP main code

The code contains:
- Loading the model
- Gathering the data
- Generating tokens representing the input
- Masking the input based on the masked tokens
- Running the model on all combinations of masked input
- Storing the predictions and probabilities

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+    args = options.parse_args_and_arch(parser)
+
+    # only override args that are explicitly given on the command line
+    override_parser = options.get_validation_parser()
+    override_args = options.parse_args_and_arch(override_parser, suppress_defaults=True)
```
Will these be used in any function call? They are unused at the moment

Kathleendeboer replied:
Removed these

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+    utils.import_user_module(args)
+    use_fp16 = args.fp16
+    use_cuda = torch.cuda.is_available() and not args.cpu
+    load_models(args)
```
Why do you load models twice?

Kathleendeboer replied:
The combination of the model was still in progress :) removed the duplicate line

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+        )
+
+
+    results_dict = {} ## This dict will gatter the shap values for each data sample key: sample ID value: shap-values for each combination of tokens
```
```suggestion
    results_dict = {} ## This dict will gather the shap values for each data sample ({key: sample ID; value: shap-values} for each combination of tokens)
```
It's a bit confusing what the format is (also, if you have multiple results, shouldn't this be an array of key-value pairs?)

Veldhoen replied:
If you need to keep somewhat complex data, you might want to consider creating a dataclass (https://docs.python.org/3/library/dataclasses.html) to make very explicit what it entails. 
Also, you can use type hints (https://docs.python.org/3/library/typing.html) to keep track of what variables are. Mypy can then automatically check if your types are correct over the flow of your code.

Kathleendeboer replied:
Thanks! I will explain @greenw0lf , the dictionary will look like: dict{
key: sample id, value: dict{key: combination id, value: prediction (label), probabilities (array of probs). 
I will test if this is a good format and otherwise look into your tips @Veldhoen!

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+        print(sample)
+        #DATA TEST, the code below is made for testing
+        #x = {'net_input': {'audio': audio_tensor, 'text': text_tensor, 'video': video_tensor, 'padded_amount':padded_amount_value }}
+        #audio_tensor = torch.randn(1, 150000)  # Example dimensions (batch_size, channels, audio_length)
+        #text_tensor = torch.randn(1, 30)  # Example dimensions (batch_size, seq_length, embedding_size)
```
Example dimensions number does not match the number of arguments for `torch.randn()` (3 variables in the comment vs. 2 in the actual code)

Kathleendeboer replied:
Thanks for pointing that out, I changed the code but not the comment..

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+        last_non_zero_index = np.max(np.nonzero(text))
+        text_length  = len(text[:last_non_zero_index + 1])
+        text_tokens_ids = torch.tensor(range(1, text_length+1)).unsqueeze(0)
+        audio = audio_tensor.squeeze(0).cpu().numpy()
+        last_index = len(audio) - np.argmax(np.flip(audio) != 1) - 1
```
Might want to move these under the next part of the code (## Audio tokens)

Kathleendeboer replied:
Yes

greenw0lf commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+        video_token_ids = torch.tensor(range(1, 257)).unsqueeze(0)
+
+        ## All tokens are concatenated together
+        All_token_ids = torch.cat((text_tokens_ids, audio_tokens_ids, video_token_ids, ), 1).unsqueeze(1)
```
```suggestion
        all_token_ids = torch.cat((text_tokens_ids, audio_tokens_ids, video_token_ids, ), 1).unsqueeze(1)
```

greenw0lf replied:
And replace in its occurrences too

greenw0lf reviewed (commented):
Also, you should look into adding github actions/workflow to your repository because I already predict that there will be errors with the line length and overall formatting of the file.

Veldhoen commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+
+def load_models(args, override_args=None):
+    if override_args is not None:
+            overrides = vars(override_args)
```
Fix indentation of this bit

Veldhoen commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+
+        ## Move models to GPU
+        for model in models:
+            if use_fp16:
```
Undefined here (make it an argument of load_models?)

Veldhoen commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+        else:
+            overrides = None
+        ## Load ensemble and extract model
+        print('| loading model(s) from {}'.format(args.path))
```
In the end, don't use print statements but a logger: in production, there is no terminal to write output to.

Veldhoen reviewed (changes requested):
Whoohoo, your first PR!

Can you make sure everything that you want reviewed is in a runnable state? For instance, I wasn't able to run sseft_shap_new.py right away (indentation errors)

I' added a few remarks here and there. I'll continue reviewing on Monday, have to call it a day now. You can make commits in the meantime of course, to respond to Dragos' recommendations, for instance :)

Veldhoen commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+            for k in range(0, len(masked_audio_tokens_ids)):
+                if audio_tokens_ids[k] == 0:
+                    masked_audio_tensor[k:k+samples_per_token] = 0 ## Here the real audio masking happens
+                        masked_video_tensor = video_tensor.clone()
```
This yields an indentation error

Veldhoen commented on sseft_shap_new.py:
```diff
@@ -0,0 +1,189 @@
+
+
+import shap
+import torch
+import numpy as np
+from PIL import Image
+import os, copy, sys
+import math, json
+import random
+from tqdm import tqdm
+from fairseq import checkpoint_utils, options, progress_bar, utils
```
Many dependencies have not been declared in pyproject.toml yet
Also, quite some imports are never used in the module
And the submodules (fairseq) have their own requiremens that are not listed in pyproject.toml

Veldhoen commented:
When I fix the missing dependencies and indentation errors and run the code, I still get a KeyError. Please try to get everything in a state that is runnable. Don't hesitate to ask for help on this though!
Let me know when you have a running version that you want a review on :)

Veldhoen reviewed (changes requested):
When I fix the missing dependencies and indentation errors and run the code, I still get a KeyError. Please try to get everything in a state that is runnable. Don't hesitate to ask for help on this though!
Let me know when you have a running version that you want a review on :)

Kathleendeboer committed:
Update sseft_shap_new.py

Debugged untill my colab resources hit their limit. It is not fully running but it comes close. Mostly i fixed an issue with the access of the original input (which i could not give as an argument, while this is not compatible with shap) I made these global.

Kathleendeboer committed:
Update dependencies

Updated dependencies in pyproject.toml and deleted imports that were not used anymore.

Kathleendeboer committed:
Added github/workflows

Kathleendeboer committed:
Deleted files not nescessary

Kathleendeboer committed:
Added torchvision

Kathleendeboer committed:
Deleted all files from sseft except for fairseq, cleaned my main code sseft_shap_github

Deleted all files from sseft except for fairseq, cleaned my main code sseft_shap_github.py

Kathleendeboer committed:
Added notebook for testing (test.ipynb)

Read README to download checkpoint, pre-trained models and test samples.
The code segments video/audio put them in correct folders and run the emotion recognition model with explainability framework.

Kathleendeboer commented:
Recently pushed running code for explainable emotion recognition with segmentation. I provided a test notebook for running an example. #2 #16

Kathleendeboer committed:
Fault in audio masking

Fault in audio masking

Veldhoen commented:
Your pyproject.toml seems to be invalid. Can you fix it to make sure your project can be installed and has all the proper dependencies? Or, at the minimum, provide a requirements.txt 
Also, I got some indentation errors upon importing your data preprocessing module: something weird seems to have happened with your docstrings.
And please indicate what files to put where in order to run your test (model checkpoint, pretrained models, input files..)

Kathleendeboer committed:
requirements + indentation

requirements + indentation

Kathleendeboer committed:
Make subdir for data

Make subdir for data

Kathleendeboer committed:
Update requirements

Update requirements

Veldhoen commented on data_preprocessing.py:
```diff
@@ -0,0 +1,122 @@
+def Segment_audio_and_video(full_audio_file, full_video_file, output_dir):
+    '''
+  Given the full audio and video file, segment the audio and video and pass them to further preprocess functions.
+  Extract transcribtions from the audio. 
```
Refactor: it would be clearer to have a separate function to transcribe the audio first, and then a separate function to segment audio and video accordingly

Veldhoen commented on data_preprocessing.py:
```diff
@@ -0,0 +1,122 @@
+
+    #--- PARAMS ---#
+    #sampling rate
+    sr = 16000
```
Parameters should in the end be configurable (typically). Although in this case, you might actually want to infer sampling rate from the audio file itself?

Veldhoen commented on data_preprocessing.py:
```diff
@@ -0,0 +1,122 @@
+        tokenized_text = Tokenize_text(name, segment, segment.text, output_dir)
+    
+        #SEGMENT AUDIO
+        seg = segment_audio(segment, audio_in, output_dir, sr, save_into_file=True)
```
This variable (and other variables in this function definition) are never accessed, so do not define them (you can use anonymous variable `_` if you need part of a result)

Veldhoen commented on fairseq/criterions/emotion_prediction_cri.py:
It's not entirely clear what the status of this file is (what is the original/what is added, for instance). 
I cannot run the entire pipeline, but my code editor complains that sample_size is never defined.

Veldhoen commented on fairseq/data/raw_audio_text_video_dataset.py:
Is this restored to original or..??

Kathleendeboer replied:
Almost, apart from: 
words.extend(line.strip().split(' ')) 
and cv2.setHeadless()

Veldhoen commented on fairseq/models/mulT_emo.py:
```diff
@@ -68,8 +68,7 @@ def __init__(self, args, encoder):
         #list_of_models =  ['/home/gsir059/Documents/Shamane-PhD/Data-Checkpoints/trained_models/Fabnet/nv2_curriculum.pth']
-        #list_of_models =  ['./trained_models/fabnet/affectnet_4views.pth']
-        list_of_models =  ['/home/kdeboer/sseft/pre-trained_models/affectnet_4views.pth'] #MODIFIED
+        list_of_models =  [os.path.join(pre_trained_path, 'affectnet_4views.pth')]
```
Ideally, model specification is configurable - not harcoded

Veldhoen commented on fairseq/models/mulT_emo.py:
```diff
@@ -50,8 +50,8 @@ def __init__(self, args, encoder):
-        #list_of_models = list_of_models =['./trained_models/wav2vec/wav2vec_large.pt'] #MODIFIED
-        list_of_models = list_of_models =['/home/kdeboer/sseft/pre-trained_models/wav2vec_large.pt']
+        pre_trained_path = args.user_dir
+        list_of_models =[os.path.join(pre_trained_path, 'wav2vec_large.pt')]
```
Ideally, model specification is configurable - not harcoded

Veldhoen commented on models.py:
```diff
@@ -0,0 +1,39 @@
+    # audio_file_path: str  # path to the audio file (.wav file)
+    # audio_tensor_path: str  # path to the audio tensor (.pt file)
+    # text_file_path: str  # text data (.txt file)
+    data_folder_path: str # folder with the video audio and text folders and the tsv files and fake label files. (to be made)
```
Make distinction between input and output - files that are created within this worker are not input to the worker

Veldhoen commented on pyproject.toml:
As discussed, the current pyproject.toml is not valid

Veldhoen commented on segment.py:
```diff
@@ -0,0 +1,115 @@
+  torchaudio.save(output_path, audio_segment, sr, format="wav")
+
+
+  return audio_segment
```
return value does not seem to be used anymore.

Veldhoen commented on segment.py:
```diff
@@ -0,0 +1,115 @@
+
+  return filename_format
+
+def segment_audio(trans_segment,
```
Are you creating both wav and pt here? I can only trace back the .wav bit - but for that, you shouldn't need torchaudio, do you?

Veldhoen commented on segment.py:
```diff
@@ -0,0 +1,115 @@
+            '''Given a waveform tensor and sample factor, resample.'''
+            return F.interpolate(x.view(1, 1, -1), scale_factor=factor).squeeze()
+
+def preprocess_audio_file(filename):
```
I think this is where the actual .pt storing happens, right?

Veldhoen commented on segment.py:
```diff
@@ -0,0 +1,115 @@
+  subprocess.call(run_ffmpeg)
+  
+
+def preprocess_data(audio_path):
```
Provide a descriptive name for this function

Veldhoen reviewed (commented):
I've mostly looked at the preprocessing modules. They seem to work! They could use a bit of refactoring here and there. And it would be nice to think about making some things configurable. 
Also, adding some docstrings would be helpful - especially for more complex functions. Lastly, please use a linter to apply flake8 and black formatting.

Kathleendeboer committed:
Handled required changes from review, mainly the data_preprocessing

Handled required changes from review, mainly the data_preprocessing and corrected pyproject.toml

Kathleendeboer committed:
Added containerisation

Changes:
- Added Dockerfile for containerisation (use GPU is required)
- Config for all input paths
I am not sure about the output of the model (it originally creates a folder where all output data is stored) but I have to specify different model outputs types in model.py. Can I just specify the output dir?

Kathleendeboer committed:
Changed Readme

new link for checkpoint

Veldhoen commented on sseft_shap.py:
```diff
@@ -0,0 +1,53 @@
+    get_base_output_dir,
+    get_output_file_path,
+)
+import SSeft_shap_github as ss
```
Import statement doesn't match module name (casing is wrong)

Veldhoen replied:
Same in structure_data.py: failing import statement because of casing

Veldhoen commented on fairseq/data/raw_audio_text_video_dataset.py:
```diff
@@ -8,6 +8,7 @@
 from .import FairseqDataset
 
 import cv2
+cv2.setHeadless() #MODIFIED
```
What should this do? It's giving me an AttributeError

Kathleendeboer committed:
Import fixes

Kathleendeboer committed:
Update readme

Veldhoen commented on Dockerfile:
```diff
@@ -1,91 +1,30 @@
+
+RUN poetry install --without dev --no-root && rm -rf $POETRY_CACHE_DIR
+# copy the rest into the source dir
+COPY ./ /src
```
Transferring the entire repository to the image is quite heavy. Instead, only include actual source - either by specifying which parts to add, or by adding a .dockerignore file that stipulates anything that doesn't need to go into the image. 
For instance, data and model folders should typically not be part of the image, but rather be mounted in the container upon running it. And a local .venv and .github typically can be left out entirely.

Veldhoen replied:
Also, add a .gitignore to make sure only relevant files end up in the repository (and no __pycache__, for instance)
For starters, you could use this default: ttps://github.com/github/gitignore/blob/main/Python.gitignore

Veldhoen commented on data_preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+
+
+
+def tsv_file(full_path,output_dir, modality, extension):
```
Function name could be a bit more descriptive

Veldhoen commented on segment.py:
```diff
@@ -0,0 +1,113 @@
+#all channels with 'downmix' 
+audio = Audio(mono='downmix') #add parsing arg (for nemo we need mono)
+
+def seconds_to_hms(seconds):
```
Redundant with data_prepocessing module

Veldhoen commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+
+Run in container with (change path for sources! and build id):
+docker run     --mount type=bind,source="/home/kdeboer/dane-emotion-recognition-worker/config",target=/root/.DANE     --mount type=bind,source="/home/kdeboer/dane-emotion-recognition-worker/data",target=/data     --mount type=bind,source="/home/kdeboer/dane-emotion-recognition-worker/Structured_input_data",target=/Structured_input_data     --mount type=bind,source="/home/kdeboer/dane-emotion-reco ...
```
Maybe use a variable for local path (like $WORKSPACE) rather than your specific local setup

Veldhoen replied:
Use the image tag (dane-emotion-recognition-worker) instead of its id (cf222214c6d3)

Veldhoen commented on Sseft_shap_github.py:
```diff
@@ -0,0 +1,497 @@
+    # never mask out CLS and SEP tokens (makes no sense for the model to work without them) -> this is handled in the load_models range(1,...) instead of range*(0,....)
+    return masked_X
+
+## This function recieves an array with rows that represent combinations of masked tokens for one sample
```
Move comments to docstrings

Veldhoen commented on Sseft_shap_github.py:
```diff
@@ -0,0 +1,497 @@
+    with torch.no_grad():
+        #Split x into 3 modalities
+        masked_text_tokens_ids = torch.tensor(x[:, :text_length]) 
+        print("masked text", masked_text_tokens_ids)
```
Use logger instead of print statements

Veldhoen commented on Sseft_shap_github.py:
```diff
@@ -0,0 +1,497 @@
+            ## Store the outcome of the model
+            probabilities = np.array(probabilities)
+            if predicted_emotion == "Neutral":
+                probability_class = probabilities[0]
```
The probability mapping should be more explicit, this is not very robust

Veldhoen commented on config/config.yml:
```diff
@@ -0,0 +1,50 @@
+    S3_BUCKET: bucket-name
+    S3_FOLDER_IN_BUCKET: folder  # folder within the bucket
+MM_EX_EMO_SETTINGS:
+    DATA_DIR: /Structured_input_data
```
Make use of the FILESYSTEM variables, rather than hardcode these locations. Structured_input_data might as well land in /data, and you can add a MODEL_DIR under FILE_SYSTEM 
This way, you only need to change one thing when running locally instead of containerized

Veldhoen commented on structure_data.py:
I'm not sure why this is a separate module. Why not add the run function to the data_preprocessing module?

Veldhoen reviewed (changes requested):
Nice work, great that you managed to somewhat wrap your work in a dane worker and containerize it. @greenw0lf , let us know whether the entire processing worked with GPU!

There are still some issues that could be improved. At the very least, make sure everything is runnable (there are some mistypes import statements) and test that this is the case by running it locally. 
Also, try to apply a linter so the automated checks don't fail. It's a good habit to use it to help you structure your code!

Commit:
Run on CPU (hacky)

greenw0lf commented:
@Veldhoen sorry for not keeping you in the loop as well! I tried to run the model via the container and using my GPU, but the conclusion that Kathleen and I arrived at is that my GPU might not have enough memory to run it (we need more than 12GB of video memory!). So, sadly we will have to find some cluster or other more powerful hardware where we can run the code

Kathleendeboer committed:
Multiple wav/mp4 as input

Multiple wav/mp4 as input, for running be sure to get the latest checkpoint from readme.

Kathleendeboer committed:
Merge branch 'MM-SHAP' of https://github.com/beeldengeluid/dane-emotion-recognition-worker into MM-SHAP

Kathleendeboer committed:
Merge check

Commit:
Simple script to obtain data, cut out relevant scenes and extract audio

greenw0lf commented:
Update: After fixing an issue I had with using the GPU in a Docker container, I tried rerunning on the GPU but it still doesn't work

Kathleendeboer committed:
Dutch textual backbone RoBERT for SSE-FT

RoBERT tokenizer in the data_preprocessing.py and visualization method in Sseft_shap_github.py. Changed the embedding dimensions from mulT_emo.py to work with the RoBERT model. Download the Dutch checkpoint from the README.md

Kathleendeboer committed:
Merge branch 'MM-SHAP' of https://github.com/beeldengeluid/dane-emotion-recognition-worker into MM-SHAP
